src/mlpack/methods/ann/layer/gru.hpp

```diff
@@ -0,0 +1,209 @@
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /*
+   * Calculate the gradient using the output delta and the input activation.
+   *
+   * @param input The input parameter used for calculating the gradient.
+   * @param error The calculated error.
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(arma::Mat<eT>&& input,
+                arma::Mat<eT>&& /* error */,
+                arma::Mat<eT>&& /* gradient */);
+
+  //! The value of the deterministic parameter.
+  bool Deterministic() const { return deterministic; }
+  //! Modify the value of the deterministic parameter.
+  bool& Deterministic() { return deterministic; }
+
+  //! Get the maximum number of steps to backpropagate through time (BPTT).
+  size_t Rho() const { return rho; }
+  //! Modify the maximum number of steps to backpropagate through time (BPTT).
+  size_t& Rho() { return rho; }
+
+  //! Get the parameters.
+  OutputDataType const& Parameters() const { return weights; }
+  //! Modify the parameters.
+  OutputDataType& Parameters() { return weights; }
+
+  //! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  //! Get the model modules.
+  std::vector<LayerTypes>& Model() { return network; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+
+  //! Locally-stored number of input units.
+  size_t inSize;
+
+  //! Locally-stored number of output units.
+  size_t outSize;
+
+  //! Number of steps to backpropagate through time (BPTT).
+  size_t rho;
+
+  //! Locally-stored weight object.
+  OutputDataType weights;
+
+  //! Locally-stored previous output.
+  arma::mat prevOutput;
+
+  //! Locally-stored input 2 gate module.
+  LayerTypes input2GateModule;
+
+  //! Locally-stored output 2 gate module.
+  LayerTypes output2GateModule;
+  
+  //! Locally-stored output hidden state 2 gate module.
+  LayerTypes outputHidden2GateModule;
+
+  //! Locally-stored input gate module.
+  LayerTypes inputGateModule;
+
+  //! Locally-stored hidden state module.
+  LayerTypes hiddenStateModule;
+
+  //! Locally-stored forget gate module.
+  LayerTypes forgetGateModule;
```

Comment:
About implementing Forward and Backward manually. I am not so sure. I did a quick and dirty check on the paper (writing down the equations involved), and did not find any operation that could be avoided with manual implementation. All the matrices that currently saved inside layers will be saved in GRU layer and they all help in Backward and Gradient. We will be saving on locality and function calls. Nonetheless, I will do the calculations again tomorrow, in case I missed out on something.